origins_workflow/origins_workflow.module

```diff
@@ -448,7 +450,77 @@ function origins_workflow_preprocess_page(&$variables) {
   // Add js to node view pages to fake a moderation
   // sidebar 'archive' button that is missing.
   $this_route = \Drupal::routeMatch()->getRouteName();
-  if ($this_route == 'entity.node.canonical') {
+  if ($this_route === 'entity.node.canonical') {
     $variables['#attached']['library'][] = 'origins_workflow/moderation_sidebar_fix';
   }
 }
+
+/**
+ * Implements hook_entity_operation().
+ */
+function origins_workflow_entity_operation(EntityInterface $entity) {
+  if ($entity instanceof ContentEntityInterface === FALSE) {
+    return;
+  }
+
+  $operations = [];
+
+  // Use the core service to fetch the permissable transitions based
+  // on underlying configuration and active user object.
+  $transitions = \Drupal::service('content_moderation.state_transition_validation')
+    ->getValidTransitions($entity, \Drupal::currentUser());
+
+  // Remove the current moderation status from the list of possible transitions.
+  // Quick publish isn't detected as an equivalent operation/status as published
+  // so we prune it off here for now as it makes no sense ot quick publish an
+  // already published piece of content.
+  if ($entity->moderation_state->value === 'published') {
+    unset($transitions['quick_publish']);
+  }
+
+  // We're going to use the permitted transitions returned above
+  // to create a collection of Url objects which we'll use to inject
+  // into the links collection. That allows editors to then select specific
+  // tasks to occur from a single location.
+
+  // Magic numbers... scheduled_transitions.module starts at 50 so a few
+  // between that and 0 (we want the 'Edit' link first) is fine.
+  $weight = 40;
+  foreach ($transitions as $id => $transition) {
+    if ($transition instanceof TransitionInterface === FALSE) {
+      // Skip over the transition if it's something other than
+      // a Transition object or something that implements that interface.
+      continue;
+    }
+
+    $url = Url::fromRoute('origins_workflow.moderation_state_controller_change_state', [
+      'nid' => $entity->id(),
+      'new_state' => $transition->to()->id(),
+    ],
+    [
+      'query' => [
+        'destination' => \Drupal::destination()->get(),
+        // Confirm query parameter triggers a message back to the user from the controller.
+        'confirm' => TRUE,
```

Comment:
There's no form POST so no way to trigger the message update, so I cheated and sneaked in a query parameter which the controller can detect and set a message there.